### ISSUE TITLE
Redefinition of `Base.eltype` for `TPS` datatype

### ISSUE THREAD

**Issue by afossa** [bmad-sim/GTPSA.jl#132]
Hello,

This is more a feature request than a bug. I would like to solve initial value problems with [`OrdinaryDiffEq`](https://github.com/SciML/OrdinaryDiffEq.jl) using `TPS` (`TPS64` in particular) as the datatype for the elements of the state vector, i.e. define and solve an `ODEProblem` with `uType Vector{TPS64}`. This is useful to compute a high-order map of the solution of the ODE with respect to variations in the initial conditions. However, I stumbled across some errors that seems all related to these definitions at lines 103-104 in [tps.jl](https://github.com/bmad-sim/GTPSA.jl/blob/main/src/tps.jl):

```julia
Base.eltype(::Type{TPS{T}}) where {T} = T
Base.eltype(::TPS{T}) where {T} = T
```

This causes some arrays being initialized with element type `Float64` when it should be `TPS64`, which result in this error being thrown when trying to fill these arrays with `TPS64` objects:

```julia
ERROR: MethodError: no method matching Float64(::TPS64)
The type `Float64` exists, but no method is defined for this combination of argument types when trying to construct it.

Closest candidates are:
  (::Type{T})(::T) where T<:Number
   @ Core boot.jl:900
  Float64(::IrrationalConstants.Log2π)
   @ IrrationalConstants ~/.julia/packages/IrrationalConstants/vp5v4/src/macro.jl:112
  Float64(::Int16)
   @ Base float.jl:239
  ...

Stacktrace:
  [1] convert(::Type{Float64}, x::TPS64)
    @ Base ./number.jl:7
  [2] setindex!(A::Vector{Float64}, x::TPS64, i::Int64)
    @ Base ./array.jl:987
  [3] macro expansion
    @ ~/.julia/packages/DiffEqBase/R2Vjs/src/calculate_residuals.jl:107 [inlined]
  [4] macro expansion
    @ ./simdloop.jl:77 [inlined]
  [5] calculate_residuals!
    @ ~/.julia/packages/DiffEqBase/R2Vjs/src/calculate_residuals.jl:106 [inlined]
  [6] perform_step!(integrator::OrdinaryDiffEqCore.ODEIntegrator{…}, cache::OrdinaryDiffEqVerner.Vern9Cache{…}, repeat_step::Bool)
    @ OrdinaryDiffEqVerner ~/.julia/packages/OrdinaryDiffEqVerner/m0ylj/src/verner_rk_perform_step.jl:1152
  [7] perform_step!
    @ ~/.julia/packages/OrdinaryDiffEqVerner/m0ylj/src/verner_rk_perform_step.jl:1043 [inlined]
  [8] solve!(integrator::OrdinaryDiffEqCore.ODEIntegrator{…})
    @ OrdinaryDiffEqCore ~/.julia/packages/OrdinaryDiffEqCore/H25Bn/src/solve.jl:569
  [9] __solve(::ODEProblem{…}, ::Vern9{…}; kwargs::@Kwargs{…})
    @ OrdinaryDiffEqCore ~/.julia/packages/OrdinaryDiffEqCore/H25Bn/src/solve.jl:7
 [10] __solve
    @ ~/.julia/packages/OrdinaryDiffEqCore/H25Bn/src/solve.jl:1 [inlined]
 [11] #solve_call#44
    @ ~/.julia/packages/DiffEqBase/R2Vjs/src/solve.jl:634 [inlined]
 [12] solve_call
    @ ~/.julia/packages/DiffEqBase/R2Vjs/src/solve.jl:591 [inlined]
 [13] #solve_up#53
    @ ~/.julia/packages/DiffEqBase/R2Vjs/src/solve.jl:1122 [inlined]
 [14] solve_up
    @ ~/.julia/packages/DiffEqBase/R2Vjs/src/solve.jl:1101 [inlined]
 [15] #solve#51
    @ ~/.julia/packages/DiffEqBase/R2Vjs/src/solve.jl:1038 [inlined]
 [16] top-level scope
    @ ~/Developer/ad_testing/lorenz63.jl:54
Some type information was truncated. Use `show(err)` to see complete types.
```

I was able to overcome this error during the solution of the initial value problem by redefining these functions in [RecursiveArrayTools.jl/blob/master/src/utils.jl](https://github.com/SciML/RecursiveArrayTools.jl/blob/master/src/utils.jl) for `AbstractVector{<:TPS}` (around line 245):

```julia
RecursiveArrayTools.recursive_bottom_eltype(v::AbstractVector{<:TPS}) = eltype(v)
RecursiveArrayTools.recursive_unitless_eltype(v::AbstractVector{<:TPS}) = eltype(v)
RecursiveArrayTools.recursive_unitless_bottom_eltype(v::AbstractVector{<:TPS}) = eltype(v)
```

I also had to define these two functions, where the second mimics what is done in [DiffEqBase.jl/blob/master/src/forwarddiff.jl](https://github.com/SciML/DiffEqBase.jl/blob/master/src/forwarddiff.jl) for `ForwardDiff` (around line 450):

```julia
Base.log10(tps::TPS) = log(tps) / log(10.0)
DiffEqBase.value(tps::TPS) = scalar(tps)
```

With these definitions I can solve the problem and extract the Taylor expansion of the solution at the time points selected by the integrator. However, if I try to interpolate the solution from the dense output returned by the solver, I obtain again the same error. This time it is thrown at line 957 of [OrdinaryDiffEqCore/src/dense/generic_dense.jl](https://github.com/SciML/OrdinaryDiffEq.jl/blob/master/lib/OrdinaryDiffEqCore/src/dense/generic_dense.jl):

```julia
out = similar(y₁, eltype(first(y₁) * oneunit(Θ)))
copyto!(out, y₁)
```

because `y₁` is (correctly) a `Vector{TPS64}` but `out` is initialized as `Vector{Float64}` since it is the datatype returned by `first(y₁)`. In contrast to the above, I do not see an easy workaround in this case.

So my question is: `GTPSA` itself seems to rely a lot on `eltype` to get the datatype of the monomial coefficients, in which case the current definition of this function for `TPS` makes perfectly sense. However, from the perspective of an external algorithm that treats `TPS` objects as "blackboxes", the "root" datatype should be `TPS64` and not `Float64`.

Is this something you would consider to change, to have `TPS` behave like `ForwardDiff.Dual` from an outside perspective? I am open to your ideas and to discuss about it. Below is the full script (with broken parts commented out) in case you want to experiment yourself.

Thank you!

```julia
using LinearAlgebra
using OrdinaryDiffEq, DiffEqBase, RecursiveArrayTools
using ForwardDiff, GTPSA
using GLMakie

import ForwardDiff: Dual

# function overloading for TPS objects
Base.log10(tps::TPS) = log(tps) / log(10.0);
DiffEqBase.value(tps::TPS) = scalar(tps);

RecursiveArrayTools.recursive_bottom_eltype(v::AbstractVector{<:TPS}) = eltype(v);
RecursiveArrayTools.recursive_unitless_eltype(v::AbstractVector{<:TPS}) = eltype(v);
RecursiveArrayTools.recursive_unitless_bottom_eltype(v::AbstractVector{<:TPS}) = eltype(v);

# dynamical model
function lorenz63!(du, u, p, t)
    σ, ρ, β = p
    du[1] = σ*(u[2] - u[1])
    du[2] = u[1]*(ρ - u[3]) - u[2]
    du[3] = u[1]*u[2] - β*u[3]
end;

# model parameters
σ = 10.0;
ρ = 28.0;
β = 8.0/3.0;

# initial conditions
u0 = [0.0, 1.0, 0.0];
Φ0 = I(3);

# propagation time span
ts = (0.0, 40.0);
ns = Int(200*ts[2]);

# differential algebra order and number of variables
no = 1;
nv = length(u0);

# solution with Float64
prob_f64 = ODEProblem(lorenz63!, u0, ts, [σ, ρ, β]);
sol_f64 = solve(prob_f64, Vern9(), reltol=1e-15, abstol=1e-15);

# solution with ForwardDiff
ode_tag = ForwardDiff.Tag{DiffEqBase.OrdinaryDiffEqTag, Dual{Nothing, Float64, nv}};
u0_fwd = [Dual{ode_tag}(u0[i], Φ0[i, :]...) for i in 1:nv];
prob_fwd = ODEProblem(lorenz63!, u0_fwd, ts, [σ, ρ, β]);
sol_fwd = solve(prob_fwd, Vern9(), reltol=1e-15, abstol=1e-15);

# solution with GTPSA
dsc = Descriptor(nv, no);
u0_tps = u0 + vars(dsc);
prob_tps = ODEProblem(lorenz63!, u0_tps, ts, [σ, ρ, β]);
sol_tps = solve(prob_tps, Vern9(), reltol=1e-15, abstol=1e-15);

# trajectory sampling
tv = range(ts[1], ts[2], length=ns);
xv_f64 = hcat(sol_f64(tv).u...);
xv_fwd = ForwardDiff.value.(hcat(sol_fwd(tv).u...));
# xv_tps = scalar.(hcat(sol_tps(tv).u...));

# final states
uf_f64 = sol_f64.u[end];
uf_fwd = ForwardDiff.value.(sol_fwd.u[end]);
uf_tps = scalar.(sol_tps.u[end]);

@assert all(isapprox.(uf_f64, uf_fwd, atol=0.5));
@assert all(isapprox.(uf_f64, uf_tps, atol=0.5));

# trajectory plot
fig = Figure();
ax = Axis3(fig[1, 1], xlabel="x", ylabel="y", zlabel="z", aspect=:equal);
lines!(ax, xv_f64[1, :], xv_f64[3, :], xv_f64[3, :], color=:blue, linewidth=2, label="Float64");
lines!(ax, xv_fwd[1, :], xv_fwd[3, :], xv_fwd[3, :], color=:red, linewidth=2, linestyle=:dash, label="ForwardDiff");
# lines!(ax, xv_tps[1, :], xv_tps[3, :], xv_tps[3, :], color=:green, linewidth=2, linestyle=:dot, label="GTPSA");
display(fig);
```

**Comment by mattsignorelli** [comment]
Thanks for submitting this! In fact, based on the `Base.eltype` definition, this is a bug:
```
  eltype(type)

  Determine the type of the elements generated by iterating a collection of the given type. For dictionary types, this
  will be a Pair{KeyType,ValType}. The definition eltype(x) = eltype(typeof(x)) is provided for convenience so that
  instances can be passed instead of types. However the form that accepts a type argument should be defined for new
  types.
```

 Coincidentally, I had just discovered this bug earlier today myself and am putting in a fix. The current (incorrect) definition for `eltype` will be changed to `numtype` in the next release to specify which type of number the `TPS` represents,  and `eltype` will be replaced with its correct definition per that in `Base`.

**Comment by mattsignorelli** [comment]
OK these issues have been fixed and will be included in the next release (1.3, hoping to go out tomorrow). I have

- Added a `log10` function
- corrected the override for `eltype`, and added a new `numtype` to get the number which the TPS represents

As for the `DiffEqBase.value(tps::TPS)` override, this is something that should belong in a GTPSA extension to `DiffEqBase`, so I am talking to the maintainers now about a PR. In the meantime , this will still have to be manually specified for integrating a `TPS` map, though.

On another note, based on your use case you might find this other package I've been developing, [NonlinearNormalForm.jl](https://github.com/bmad-sim/NonlinearNormalForm.jl), interesting. It is still early in development and I have some code cleaning to do, but all the low-level functionality is there. I am happy to hear functionality suggestions for it too.

Thanks again for submitting the issue!
